packages/loader/container-loader/src/deltaManager.ts

```diff
@@ -235,15 +235,26 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
     public submitSignal(content: any) { return this.connectionManager.submitSignal(content); }
 
     public flush() {
-        if (this.messageBuffer.length === 0) {
+        const batch = this.messageBuffer;
+        if (batch.length === 0) {
             return;
         }
 
+        this.messageBuffer = [];
+
         // The prepareFlush event allows listeners to append metadata to the batch prior to submission.
-        this.emit("prepareSend", this.messageBuffer);
+        this.emit("prepareSend", batch);
 
-        this.connectionManager.sendMessages(this.messageBuffer);
-        this.messageBuffer = [];
+        if (batch.length === 1) {
+            assert(batch[0].metadata?.batch === undefined, "no batch markup on single message");
+        } else {
+            assert(batch[0].metadata?.batch === true, "no start batch markup");
+            assert(batch[batch.length - 1].metadata?.batch === false, "no end batch markup");
+        }
+
+        this.connectionManager.sendMessages(batch);
+
+        assert(this.messageBuffer.length === 0, "reentrancy");
```

Comment:
I'm not sure what this assert is guarding against. I don't think there's a way for the buffer to have new items at the end of this function as everything is synchronous and flush can't be called in the meantime. Is this to future-proof the mechanics of it?

Reply:
"prepareSend" event can be the source of reentrancy. Yes, chances are low, but when these issues happen, it's pain in the butt to diagnose them, so I prefer defensive programming :)